src/client/Client.cc

```diff
@@ -11029,6 +11029,15 @@ void Client::do_readahead(Fh *f, Inode *in, uint64_t off, uint64_t len)
 
 void Client::C_Read_Async_Finisher::finish(int r)
 {
+  // If inode's file cache cap ref count is more than one then it asserts
+  // ObjectCacher::file_read for async IO returned 0 and file cache cap ref
+  // was incremented. So, it is a must to put it down before readahead
+  // otherwise it will stall the IO call since it will keep on waiting for
+  // the file cache cap to be released.
+  if ((in->cap_refs.at(CEPH_CAP_FILE_CACHE)) > 1) {
+      clnt->put_cap_ref(in, CEPH_CAP_FILE_CACHE);
+  }
+
```

Comment:
it isnt buggy, you need Fc cap while waiting for IO to finish. Take a look at the sync IO code:

```
  if (r == 0) {
    get_cap_ref(in, CEPH_CAP_FILE_CACHE);
    client_lock.unlock();
    r = io_finish_cond->wait();
    client_lock.lock();
    put_cap_ref(in, CEPH_CAP_FILE_CACHE);
    update_read_io_size(bl->length());
  }
```

The same needs to be done for async IO, so the put_cap_ref part needs to be done in finish().

Reply:
No, I just mean there are two places will call the `do_readahead()`, but they are doing differently. One will just try to take the ref for `Fc` caps, but another won't.

Reply:
so one place is for sync io and other one is for async io. The do_readahead() call in Client::_read_async is for sync io since the async call returns 0 before it reaches that line

Reply:
> but they are doing differently. One will just try to take the ref for `Fc` caps, but another won't.

actually both do the readahead with Fc cap, its just that sync call execution is straightforward while for async io it will get the cap ref in the else part:

```
  if (onfinish != nullptr) {
    // Release C_Read_Async_Finisher from managed pointer, either
    // file_read will result in non-blocking complete, or we need to complete
    // immediately. In either case, the C_Read_Async_Finisher is safely
    // handled and won't be abandoned.
    Context *crf = io_finish.release();
    if (r != 0) {
      // need to do readahead, so complete the crf
      crf->complete(r);
    } else {
      get_cap_ref(in, CEPH_CAP_FILE_CACHE);
    }
    return 0;
  }
```

Reply:
after this the call reaches `Client::C_Read_Async_Finisher::finish` and does the readahead. So basically there are going the same thing but async code path is a bit spread over the places

Reply:
Here let's just ignore the `CEPH_CAP_FILE_LAZYIO` case.

So when the `_read_async()` is called that means the `Fc` caps has already held by the caller.

While in case `_read_async()-->do_readahead()` it doesn't try to get the `Fc` caps. But in async case it just double get the 'Fc' caps and then your fix in this PR will try to release it, right ?